sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -474,203 +641,234 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
-            stateDecl != null,
-            "%s parameter of type %s at index %s references undeclared %s \"%s\"",
-            fnClass.getRawType().getName(),
-            params[i],
-            i,
-            DoFn.StateId.class.getSimpleName(),
-            id);
+      TimerDeclaration timerDecl = globalContext.getTimerDeclarations().get(id);
+      errors.checkArgument(
+          timerDecl != null,
+          "parameter of type %s at index %s references undeclared %s \"%s\"",
+          param.getType(),
+          param.getIndex(),
+          TimerId.class.getSimpleName(),
+          id);
 
-        errors.checkArgument(
-            stateDecl.stateType().equals(stateType),
-            "%s parameter at index %s has type %s but is a reference to StateId %s of type %s",
-            fnClass.getRawType().getName(),
-            i,
-            params[i],
-            id,
-            stateDecl.stateType());
+      errors.checkArgument(
+          timerDecl.field().getDeclaringClass().equals(param.getMethod().getDeclaringClass()),
+          "parameter at index %s for timer %s"
+              + " declared in a different class %s."
+              + " Timers may be referenced only in the lexical scope where they are declared.",
+          Timer.class.getSimpleName(),
+          param.getIndex(),
+          id,
+          timerDecl.field().getDeclaringClass().getName());
 
-        errors.checkArgument(
-            stateDecl.field().getDeclaringClass().equals(m.getDeclaringClass()),
-            "Method %s has State parameter at index %s for state %s"
-                + " declared in a different class %s."
-                + " State may be referenced only in the class where it is declared.",
-            m,
-            i,
-            id,
-            stateDecl.field().getDeclaringClass().getName());
-
-        DoFnSignature.Parameter.StateParameter stateParameter = Parameter.stateParameter(stateDecl);
-        stateParameters.put(id, stateParameter);
-        extraParameters.add(stateParameter);
-      } else {
-        List<String> allowedParamTypes =
-            Arrays.asList(
-                formatType(new TypeToken<BoundedWindow>() {}),
-                formatType(new TypeToken<RestrictionTracker<?>>() {}));
-        errors.throwIllegalArgument(
-            "%s is not a valid context parameter. Should be one of %s",
-            formatType(paramT), allowedParamTypes);
+      return Parameter.timerParameter(timerDecl);
+
+    } else if (RestrictionTracker.class.isAssignableFrom(rawType)) {
+      errors.checkArgument(
+          !methodContext.getExtraParameters().contains(Parameter.restrictionTracker()),
+          "Multiple %s parameters",
+          RestrictionTracker.class.getSimpleName());
+      return Parameter.restrictionTracker();
+    } else if (State.class.isAssignableFrom(rawType)) {
+      // m.getParameters() is not available until Java 8
+      String id = getStateId(param.getAnnotations());
+      errors.checkArgument(
```

Comment:
Done. Kept method-level and param-level reporting btw, depending on the error, rather than trying to rephrase them all to fit one particular context.